theano/sandbox/cuda/dnn.py

```diff
@@ -2186,6 +2186,460 @@ def method(self):
 #endif
         """
 
+class GpuDnnBatchNormBase(DnnBase):
+    """
+    Op for the cuDNN Batch Normalization.
+
+    Parameters
+    ----------
+    tensor_format
+        Always set this to 'bc01'.
+    mode
+        Normalization is performed per-activation for 'per-activation' (in this mode bnBias and bnScale tensor
+        dimensions are 1xCxHxW for 4d tensor) or Normalization is performed over N+spatial dimensions for 'spatial'
+        (in this mode bnBias, bnScale tensor dimensions are 1xCx1x1).
+    exp_avg_factor
+        Factor used in the moving average computation runningMean = newMean*factor + runningMean*(1-factor). It is
+        required that exponentialAverageFactor=1 is used for the very first call of a complete training cycle.
```

Comment:
It's something that you can change between every batch, to turn the exponential averaging into an overall average. If we want to support that, it needs to be an input, yes.

The comment about setting it to 1 for the first call (not for the first epoch!) is generally valid, but I don't think it's relevant in practice. Usually you'd just initialize the runningMean to zero and the runningInvStd to one and then use the same factor throughout training. The running averages will only be used for validation, after a full epoch of training, so the initialization will have been mostly "forgotten" by then (e.g., only included with a factor of 0.9^num_training_batches).